### PR TITLE
chore(typescript): migrate protractor.js to TypeScript.

### PR DESCRIPTION
**Pull request by mgiambalvo**
This isn't done yet, but it's a big one so I'd like some early feedback.

### CODE REVIEW

**Comment by mgiambalvo** [comment]
Oh and I'm pretty sure I broke a couple things in the process, so I need to fix tests as well.

File: lib/globals.d.ts [hunk 1]
```diff
@@ -1,12 +1,28 @@
...
+    navigate: Function;
+    get: Function;
+    wait: Function;
+    schedule: Function;
+    controlFlow: Function;
+    findElements: Function;
     static attachToSession: Function;
   }
 
+  class Locator {
```
Comment (juliemr) [inline lib/globals.d.ts]: does it make sense to get the locator file migrated before this one?

File: lib/protractor.ts [hunk 2]
```diff
@@ -0,0 +1,1208 @@
...
+ *
+ * @private
+ * @param {!(string|Function)} script The script to execute.
+ * @param {string} description A description of the command for debugging.
+ * @param {...*} var_args The arguments to pass to the script.
+ * @return {!webdriver.promise.Promise.<T>} A promise that will resolve to the
+ *    scripts return value.
+ * @template T
+ */
+  // TODO add varargs parameter
```
Comment (juliemr) [inline lib/protractor.ts]: Can remove this todo now, right?

File: lib/protractor.ts [hunk 3]
```diff
@@ -0,0 +1,1208 @@
...
+   * });
+   *
+   * @param {!string} name The name of the module to load or override.
+   * @param {!string|Function} script The JavaScript to load the module.
+   *     Note that this will be executed in the browser context, so it cannot
+   *     access variables from outside its scope.
+   * @param {...*} varArgs Any additional arguments will be provided to
+   *     the script and may be referenced using the `arguments` object.
+   */
+  addMockModule(name: string, script: string|Function, ...moduleArgs: Array<any>) {
```
Comment (juliemr) [inline lib/protractor.ts]: Prefer `any[]` to `Array<any>` here and throughout.

File: lib/protractor.ts [hunk 4]
```diff
@@ -0,0 +1,1208 @@
...
+ * @private
+ * @param {!(string|Function)} script The script to execute.
+ * @param {string} description A description for debugging purposes.
+ * @param {...*} var_args The arguments to pass to the script.
+ * @return {!webdriver.promise.Promise.<T>} A promise that will resolve to the
+ *    scripts return value.
+ * @template T
+ */
+  // TODO add varargs parameter
+  private executeAsyncScript_(script: string|Function, description: string):
```
Comment (juliemr) [inline lib/protractor.ts]: I think we need to add another `...args` here

File: lib/protractor.ts [hunk 5]
```diff
@@ -0,0 +1,1208 @@
...
+      ...args: Array<any>): webdriver.Promise {
+    if (typeof script === 'function') {
+      script = 'return (' + script + ').apply(null, arguments);';
+    }
+
+    // TODO use args instead of arguments
+    return this.driver.schedule(
+        new Command(CommandName.EXECUTE_SCRIPT)
+            .setParameter('script', script)
+            .setParameter('args', Array.prototype.slice.call(arguments, 2)),
```
Comment (juliemr) [inline lib/protractor.ts]: This can be `args` instead of slicing `arguments`

File: lib/protractor.ts [hunk 6]
```diff
@@ -0,0 +1,1208 @@
...
+
+/**
+ * @type {ProtractorBy}
+ */
+export let By = new ProtractorBy();
+
+/**
+ * @type {ElementFinder}
+ */
+export {ElementFinder} from './element';
```
Comment (juliemr) [inline lib/protractor.ts]: I think we can do `export {ElementFinder, ElementArrayFinder} from './element';`

**Comment by juliemr** [comment]
Some early comments - overall strategy looks OK once we get the wrapping working.

File: lib/globals.d.ts (hunk 1 above)
Comment (mgiambalvo) [inline lib/globals.d.ts]: That might be helpful.

File: lib/protractor.ts (hunk 6 above)
Comment (mgiambalvo) [inline lib/protractor.ts]: What should the JSDoc be like if we did that? Or does it matter?

Comment (juliemr) [inline lib/protractor.ts]: I don't _think_ we use these parsed types in our docs.

**Comment by mgiambalvo** [comment]
Please take a look. This is mostly ready, though I still need to fix the interactive tests.

**Comment by juliemr** [comment]
LGTM once things are passing.

**Pull request merged**